File: java/src/main/java/ls.java
import picocli.CommandLine;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;

import java.io.IOException;
import java.nio.file.DirectoryStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Comparator;
import java.util.stream.StreamSupport;

@Command(name="ls", description = "List directory contents")
public class ls implements Runnable{

    @Option(names = {"-d", "--directory"}, defaultValue = ".", description = "Specify the directory to list")
    private String directory;

    @Option(names = {"-s", "--sort"}, description = "Sort the output alphabetically in ascending order")
    private boolean sort;
    @Option(names = {"-r", "--reverse"}, description = "Sort the output alphabetically in descending order")
    private boolean reverse;
    @Option(names = {"-a", "--all"}, description = "Show all files, including hidden files")
    private boolean all;
    @Option(names = {"-f", "--files"}, description = "Show only regular files")
    private boolean files;
    @Option(names = {"-h", "--help"}, usageHelp = true, description = "Show this help message")
    private boolean helpRequested;

    @Override
    public void run() {
        if (helpRequested) {
            CommandLine.usage(this, System.out);
        }
        Path path = Paths.get(directory);
        try (DirectoryStream<Path> stream = Files.newDirectoryStream(path)) {
            var fileStream = StreamSupport.stream(stream.spliterator(), false)
                    .filter(p -> {
                        return all || !p.toFile().isHidden();
                    }).sorted();
            if (reverse) {
                fileStream = fileStream.sorted(Comparator.reverseOrder());
            }
            fileStream.forEach(
                    p -> System.out.println(p.getFileName()));
        } catch (IOException e) {
            System.err.println("No such directory: " + path);
        } catch (SecurityException e) {
            System.err.println("Permission denied");
        }
    } 

    public static void main(String[] args) {
        ls cli = new ls();
        new CommandLine(cli).execute(args);
        

    }
}


    

File: go/cmd/ls.go
/*
Copyright © 2025 Harrison Jansen van Beek <harrison@jansenvanbeek.me>
*/
package cmd

import (
	"fmt"
	"log"
	"os"
	"sort"
	"strings"
	"github.com/spf13/cobra"
)
var dir *string
var shouldSort, reverse, all, files *bool

// lsCmd represents the ls command
var lsCmd = &cobra.Command{
	Use:   "ls",
	Short: "list all entries in directory",
	Long: `This function mimics the functionality of the unix ls command. **Add more detail here.`,
	Run: func(cmd *cobra.Command, args []string) {
		entries, err := os.ReadDir(*dir)
		if err != nil {
			log.Fatal(err)
		}
		if *reverse {
			sortFiles(&entries)
		}
		for _, entry := range entries {
			if !*all && strings.HasPrefix(entry.Name(), "."){
				continue
			}
			fmt.Println(entry.Name())

		}

	},
}

func init() {

	dir = lsCmd.PersistentFlags().StringP("directory", "d", ".", "set a directory to read all entries in")
	shouldSort = lsCmd.PersistentFlags().BoolP("sort", "s", true, "Sort output alphabetically in ascending order")
	reverse = lsCmd.PersistentFlags().BoolP("reverse", "r", false, "Sort output alphabetically in descending order")
	all = lsCmd.PersistentFlags().BoolP("all", "a", false, "show hidden files in output")
	files = lsCmd.PersistentFlags().BoolP("files", "f", true, "show only regular files in output")

	rootCmd.AddCommand(lsCmd)

	// Here you will define your flags and configuration settings.

	// Cobra supports Persistent Flags which will work for this command
	// and all subcommands, e.g.:
	// lsCmd.PersistentFlags().String("foo", "", "A help for foo")

	// Cobra supports local flags which will only run when this command
	// is called directly, e.g.:
	// lsCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
}

func sortFiles(entries *[]os.DirEntry) {
	if entries == nil || len(*entries) == 0 {
        return
    }
    s := *entries
    sort.SliceStable(s, func(i, j int) bool {
            return s[i].Name() > s[j].Name()
    })
    *entries = s

}
